### ISSUE TITLE
Header includes specified as preprocessor macros are not resolved

### ISSUE THREAD

**Issue by cdwilson** [microsoft/vscode-makefile-tools#196]
The [micropython](https://github.com/micropython/micropython) Makefile specifies some of the header includes as preprocessor macros.  For example, `STM32_HAL_H` is defined in `ports/stm32/Makefile` as:

```
CFLAGS += -DSTM32_HAL_H='<stm32$(MCU_SERIES)xx_hal.h>'
```

This shows up as an include in the source files like this:

```C
// ports/stm32/mphalport.h
#include STM32_HAL_H
```

Makefile Tools doesn't seem to be able to resolve headers defined this way.  If I replace the preprocessor macro with the actual header file, the red squiggles go a way and IntelliSense starts working:

```diff
-#include STM32_HAL_H
+#include <stm32f4xx_hal.h>
```

**Comment by andreeis** [comment]
@cdwilson, thank you for describing this scenario. We weren't aware of it. It's very interesting and we will address this case.

**Comment by andreeis** [comment]
Also, can you share your .vscode/dryrun.log for this project when you don't change the header definition?
And define "makefile.extensionLog": ".vscode/extension.log" and "makefile.loggingLevel": "Debug" and share .vscode/extension.log.

**Comment by cdwilson** [comment]
@andreeis please see attached

`settings.json`
```json
{
    "C_Cpp.default.configurationProvider": "ms-vscode.makefile-tools",
    "makefile.extensionOutputFolder": "./.vscode",
    "makefile.configurations": [
        {
            "name": "PYBv1.1",
            "makeArgs": [
                "BOARD=PYBV11",
                "V=1"
            ]
        }
    ],
    "makefile.phonyOnlyTargets": true,
    "makefile.loggingLevel": "Debug",
    "makefile.extensionLog": ".vscode/extension.log",
    "makefile.makeDirectory": "./ports/stm32/"
}
```

[dryrun.log](https://github.com/microsoft/vscode-makefile-tools/files/6773495/dryrun.log)
[extension.log](https://github.com/microsoft/vscode-makefile-tools/files/6773496/extension.log)

![image](https://user-images.githubusercontent.com/14631/124683558-24d53780-de82-11eb-8675-d41767b73bca.png)

**Comment by TomIowaSoft** [comment]
In my case Intellisense never learns of any of the macros defined via the compiler -D argument,  when building with gcc.

**Comment by andreeis** [comment]
@TomIowaSoft, when we will get to work on a fix for this problem we will let you know and give you a private vsix to install and confirm if you are unblocked for your specific scenario.
If your -D switch is not of the type described above at the beginning (like CFLAGS += -DSTM32_HAL_H='<stm32$(MCU_SERIES)xx_hal.h>') then let's create a new issue.

**Comment by TomIowaSoft** [comment]
My -D switches are specified in APP_DEFINES rather than added to CFLAGS.

**Comment by andreeis** [comment]
Oh ok, if that is the only difference then your scenario should be the same and we will address it with the same fix.

**Comment by andreeis** [comment]
@cdwilson, @TomIowaSoft, can you cut the quotes and define like this?
CFLAGS += -DSTM32_HAL_H=<stm32$(MCU_SERIES)xx_hal.h>

I tried this pattern and without quotes I get a valid include. Also, make sure you have Makefile Tools 0.3.0 which we released last week. Maybe with the previous 0.2.2 you were not able to use this define without quotes because of some other bugs.

**Comment by andreeis** [comment]
@cdwilson, thank you for testing the new release and we are looking forward for your feedback.
Regarding the quotes, thanks for explaining where those come from. I will revisit my investigation then.

**Comment by TomIowaSoft** [comment]
I am unable to build anything with current makefile tools. I get this. It appears to not like the extra set of ' ' around all the arguments to make.  If I copy the command line and remove the extra ' ' around the paths then it works when I enter it in the terminal.

Executing task: 'c:\program files\git\usr\bin\make.exe' 'all' 'mcu=1' 'build=debug' '-f' '"d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu\makefile"' '-C' '"d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu"' <

make: *** "d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu": No such file or directory.  Stop.
The terminal process "C:\Program Files\Git\bin\bash.exe '--login', '-c', ''c:\program files\git\usr\bin\make.exe' 'all' 'mcu=1' 'build=debug' '-f' '"d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu\makefile"' '-C' '"d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu"''" terminated with exit code: 2.

**Comment by TomIowaSoft** [comment]
Also, my makefile does not appear to have any ' ' or " " in any of the /D for mcu=1. It uses " " for /I

**Comment by TomIowaSoft** [comment]
Looking at the dryrun.log it shows this command line at the top.
c:\program files\git\usr\bin\make.exe all mcu=2 build=debug -f "d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu\makefile" -C "d:\OneDrop\sano-potentiostats-firmware\wpstat\mcu" --dry-run --always-make --keep-going --print-directory -e mcu=1

Where does -e mcu=1 come from since I specified mcu=2?

**Comment by TomIowaSoft** [comment]
This command works if I enter it in the terminal.

**Comment by andreeis** [comment]
@TomIowaSoft, thank you for the comments so far, this is very valuable. I am investigating all of these scenarios and will be back with an update.

**Comment by TomIowaSoft** [comment]
The configurationCache.log appears to be parsing the quoted values correctly. ie: "USB_MANUFACTURER='\"Sano mcu2\"' although Intellisense does not realize it.  

In the code it needs to have the double quotes in the value of USB_MANUFACTURER.

**Comment by andreeis** [comment]
@cdwilson and @TomIowaSoft, if you manually install the vsix from [this link](https://github.com/microsoft/vscode-makefile-tools/actions/runs/1556708172) until we release 0.3.1, you can be unblocked with the quoting build problem.
Continuing to work on the other issues, quotes for -D define switches, etc... and anything else was reported.
Let me know how the new vsix works for you.

**Comment by andreeis** [comment]
@TomIowaSoft, I went through all your other comments.

Regarding the USB_MANUFACTURER define: don't change that " into <. I suggested such change only for defines that in the end are used for #include paths in c/c++ files. Like the first pattern in this bug report. Your second define scenario is different and should work in its original form. How is  '-DUSB_MANUFACTURER="Sano $(TARGET)"' working for you now?

Regarding the first pattern, I may have a fix for that, even if I said that I recommended the define to be changed. Since it's from a third party, maybe it's a valid scenario we should support. I'll be back with update on this.

Regarding mcu 1 and 2: can I see your settings.json? I am curious of "makefile.configurations" and which one you chose in the left Makefile Tools UI panel ("configuration" item)? We construct that command line from 2 sources: the "makefile.configurations" entry corresponding to the "configuration" item in the Makefile Tools left side panel + "makefile.additionalDryRunSwitches" array (did you happen to leave by mistake -e mcu=1 in there maybe?).

**Comment by TomIowaSoft** [comment]
It is better than it was but some -Ddefine do not appear in Intellisense.  For example, for the following, SANO_USB_DELAY has a value when I hover over it in the .cpp file, but USBCON does not. However the #ifdef USBCON seems to have evaluated to true so it must have a value. Perhaps it is a quirky bug in Intellisense.

"C:/Tools/GNU Tools ARM Embedded/7 2017-q4-major/bin/arm-none-eabi-g++" -c -DSANO -Wall  -Og -ggdb3 -MMD  -std=gnu++11 -ffunction-sections -fdata-sections -nostdlib -fno-threadsafe-statics -fno-rtti -fno-exceptions  --param max-inline-insns-single=500  -mcpu=cortex-m0plus -mthumb -Werror -DDEBUG -DJORDAN -DSANO_WPSTAT -DWPSTAT -DMAKEFILE_BUILD -DARM_MATH_CM3 -DF_CPU=48000000L -DARDUINO=10805 -DARDUINO_SAM_ZERO -DARDUINO_ARCH_SAMD -D__SAMD21G18A__ -DUSE_32K_CLOCK_INPUT -DUSE_USB_SERIAL -Dprintf=iprintf -DMCU_NUMBER=2 -DUSBCON=1  -DSANO_USB_DELAY=500 -DUSB_CONFIG_POWER=2
-DUSB_VID=0x2a03 -DUSB_PID=0x804f -DUSB_MANUFACTURER='"Sano mcu2"' -DUSB_PRODUCT='"wpstat mcu2"' -I"." -I"./.." -I"./libraries/TaskScheduler" -I"../arduino/hardware/sano_samd/variants/sano_wpstat_mcu2_5" -I"../arduino/hardware/sano_samd/cores/arduino" -I"../arduino/hardware/sano_samd/libraries/SPI/src" -I"../arduino/hardware/sano_samd/libraries/Wire/src" -I"./libraries/ZeroTimer" -I"../arduino/hardware/sano_samd/system/CMSIS/CMSIS/Include" -I"../arduino/hardware/sano_samd/system/CMSIS/Device/ATMEL"  -I"../arduino/tools/CMSIS-Atmel/1.1.0/CMSIS/Device/ATMEL" -I"../arduino/tools/CMSIS/4.5.0/CMSIS/Include" arduino_main.cpp -o _debug_mcu2/arduino_main.cpp.o

**Comment by TomIowaSoft** [comment]
This code it got wrong in intellisense.

**Comment by TomIowaSoft** [comment]
It looks like things maybe working correctly now. Thank you.  Best support I have ever had!

**Comment by andreeis** [comment]
@TomIowaSoft, very glad to hear this. If you would like, you can also leave us a review on [Microsoft Market Place](https://marketplace.visualstudio.com/items?itemName=ms-vscode.makefile-tools&ssr=false#review-details).

**Comment by gcampbell-msft** [comment]
Closing this based on https://github.com/microsoft/vscode-makefile-tools/issues/196#issuecomment-990347488.

Thanks!
